securesystemslib/interface.py

```diff
@@ -229,187 +247,191 @@ def generate_and_write_rsa_keypair(filepath=None, bits=DEFAULT_RSA_KEY_BITS,
+  The expected key format is PKCS#1. If a password is passed or entered on the
+  prompt, the private key is decrypted, otherwise it is treated as unencrypted.
 
-    # TODO: PASSWORD_SCHEMA should be securesystemslib.schema.AnyString(min=1)
-    if not len(password):
-      raise ValueError('Password must be 1 or more characters')
-
-  elif prompt:
-    # Password confirmation disabled here, which should ideally happen only
-    # when creating encrypted key files (i.e., improve usability).
-    # It is safe to specify the full path of 'filepath' in the prompt and not
-    # worry about leaking sensitive information about the key's location.
-    # However, care should be taken when including the full path in exceptions
-    # and log files.
-    # NOTE: A user who gets prompted for a password, can only signal that the
-    # key is not encrypted by entering no password in the prompt, as opposed
-    # to a programmer who can call the function with or without a 'password'.
-    # Hence, we treat an empty password here, as if no 'password' was passed.
-    password = get_password('Enter a password for an encrypted RSA'
-        ' file \'' + TERM_RED + filepath + TERM_RESET + '\': ',
-        confirm=False) or None
+  Arguments:
+    filepath: The path to read the file from.
+    password (optional): A password to decrypt the key.
+    scheme (optional): The signing scheme assigned to the returned key object.
+        See RSA_SCHEME_SCHEMA for available signing schemes.
+    prompt (optional): A boolean indicating if the user should be prompted
+        for a decryption password. If the user enters an empty password, the
+        key is not decrypted.
+    storage_backend (optional): An object implementing StorageBackendInterface.
+        If not passed a default FilesystemBackend will be used.
 
-  if password is not None:
-    # This check will not fail, because a mal-formatted passed password fails
-    # above and an entered password will always be a string (see get_password)
-    # However, we include it in case PASSWORD_SCHEMA or get_password changes.
-    securesystemslib.formats.PASSWORD_SCHEMA.check_match(password)
+  Raises:
+    UnsupportedLibraryError: pyca/cryptography is not available.
+    FormatError: Arguments are malformed.
+    ValueError: Both a 'password' is passed and 'prompt' is true.
+    StorageError: Key file cannot be read.
+    CryptoError: Key cannot be parsed.
 
-  else:
-    logger.debug('No password was given. Attempting to import an'
-        ' unencrypted file.')
+  Returns:
+    An RSA private key object conformant with 'RSAKEY_SCHEMA'.
+
+  """
+  securesystemslib.formats.PATH_SCHEMA.check_match(filepath)
+  securesystemslib.formats.RSA_SCHEME_SCHEMA.check_match(scheme)
+
+  password = _get_key_file_decryption_password(password, prompt, filepath)
 
   if storage_backend is None:
     storage_backend = securesystemslib.storage.FilesystemBackend()
```

Comment:
Off-topic, but: now that I see how we keep using this, we should make a singleton out of `FilesystemBackend()` rather than instantiating it over and over again wherever we need to use it. @joshuagl

Reply:
Good idea! Mind submitting a feature request?